src/libraries/System.Memory.Data/src/System/BinaryData.cs

```diff
@@ -161,17 +250,29 @@ public static BinaryData FromStream(Stream stream)
         /// <param name="cancellationToken">A token that may be used to cancel the operation.</param>
         /// <returns>A value representing all of the data remaining in <paramref name="stream"/>.</returns>
         public static Task<BinaryData> FromStreamAsync(Stream stream, CancellationToken cancellationToken = default)
+            => FromStreamAsync(stream, async: true, cancellationToken: cancellationToken);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from the specified stream
+        /// and sets <see cref="MediaType"/> to <see pref="mediaType"/> value.
+        /// The stream is not disposed by this method.
+        /// </summary>
+        /// <param name="stream">Stream containing the data.</param>
+        /// <param name="mediaType">MIME type of this data, e.g. <see cref="MediaTypeNames.Application.Octet"/>.</param>
+        /// <param name="cancellationToken">A token that may be used to cancel the operation.</param>
+        /// <returns>A value representing all of the data remaining in <paramref name="stream"/>.</returns>
+        /// <seealso cref="MediaTypeNames"/>
+        public static Task<BinaryData> FromStreamAsync(Stream stream, string? mediaType, CancellationToken cancellationToken = default)
+            => FromStreamAsync(stream, async: true, mediaType, cancellationToken);
+
+        private static async Task<BinaryData> FromStreamAsync(Stream stream, bool async,
```

Comment:
This is changing the behavior of how null streams are handled.  Previously the resulting ArgumentNullException would propagate out of async calls synchronously, which is what we want to happen for usage errors.  With this change, that ArgumentNullException is now going to be stored into the returned Task.  This should also be reverted.